### DISCUSSION TITLE
4.29 needs TRUSTED_PROXY_IP set

### DISCUSSION THREAD

**Discussion by martinmuc** [mastodon/mastodon#30522] [Q&A]
Hello,
before i upgrade to 4.2.9 i want to confirm, how to handle the TRUSTED_PROXY_IP parameter.

i have several webservers and several Sidekiq Servers running.

Shall i set TRUSTED_PROXY_IP all webserver/sidekiq Server IPs on all servers, or only the own IP from the server itself?

Next question is, because it is not well documented, how do set several IPs correctly.
With a blank as separator, or a comma(,) as separator?

Thanks in advance,

Martin

**Comment by unbelauscht** [comment]
It's a comma separated list of CIDRs. You should set it for every service running behind your reverse proxy.

**Comment by letydoesstuff** [comment]
Previous reply basically nails it, but just to be explicit about how it looks in practice, you're adding something that starts with... 

```TRUSTED_PROXY_IP=127.0.0.1/8,::1/128,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16,fc00::/7```

...to `.env.production`, the above being how you'd define the default trusted IPs provided in the [documentation](https://docs-p.joinmastodon.org/admin/config/#trusted_proxy_ip). You'd then add every IP you have acting as a reverse proxy.

**Comment by racingmars** [comment]
I've upgraded from 4.2.8 to 4.2.9 and this TRUSTED_PROXY_IP thing isn't working for me. I have an internet-facing proxy server, which forwards to the ngingx instance on my Mastodon server, which forwards to Mastodon. The path from Mastodon's perspective looks like [client IP] -> [192.168.42.43] -> [127.0.0.1] -> [mastodon]

In my .env.production, I have `TRUSTED_PROXY_IP=192.168.64.32,127.0.0.1`

However, all requests to Mastodon appear to be broken, with the following in my /var/log/daemon.log:

```
Jun  6 10:31:57 mastodon bundle[89011]: F, [2024-06-06T10:31:57.620071 #89011] FATAL -- : [3d593bf1-a747-4382-acfb-4aba7e9d56e4]
Jun  6 10:31:57 mastodon bundle[89011]: [3d593bf1-a747-4382-acfb-4aba7e9d56e4] ActionDispatch::RemoteIp::IpSpoofAttackError (IP spoofing attack?! client 127.0.0.1 is not a trusted proxy HTTP_CLIENT_IP=nil HTTP_X_FORWARDED_FOR="195.201.25.153, 192.168.64.32"):
Jun  6 10:31:57 mastodon bundle[89011]: [3d593bf1-a747-4382-acfb-4aba7e9d56e4]
Jun  6 10:31:57 mastodon bundle[89011]: [3d593bf1-a747-4382-acfb-4aba7e9d56e4] lib/action_dispatch/remote_ip_extensions.rb:49:in `calculate_ip'
Jun  6 10:31:57 mastodon bundle[89011]: [3d593bf1-a747-4382-acfb-4aba7e9d56e4] config/initializers/rack_attack.rb:17:in `remote_ip'
Jun  6 10:31:57 mastodon bundle[89011]: [3d593bf1-a747-4382-acfb-4aba7e9d56e4] config/initializers/rack_attack.rb:66:in `block in <class:Attack>'
```

In the above example, 195.201.25.153 is some IP on the internet, 192.168.64.32 is the first level of proxy server, and 127.0.0.1 is presumably nginx on the mastodon server itself forwarding into mastodon.

Before upgrading to 4.2.9, the client IP address show on my account in the Mastodon web UI was correct, so it was correctly walking back up through the X-Forwarded-For chain before the upgrade.

I've tried adding the entire default list in explicitly (`TRUSTED_PROXY_IP=127.0.0.1/8,::1/128,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16,fc00::/7`), which should cover both the 127.0.0.1 proxy and any 192.168.x.x proxy, but that's no different than specifically adding 127.0.0.1 and 192.168.64.32.

Downgrading to 4.2.8 makes everything work properly again, without needing any TRUSTED_PROXY_IP entry at all.

Any hints? Both of my proxy addresses should be in the default list without me needing to even add TRUSTED_PROXY_IP, but it seems like maybe something is broken when more than one proxy server is involved now?
